File: src/main/java/com/hrms/hr_payroll_management_system/service/AttendanceService.java
package com.hrms.hr_payroll_management_system.service;

import com.hrms.hr_payroll_management_system.dto.request.attendance.AdjustAttendanceRequest;
import com.hrms.hr_payroll_management_system.dto.response.attendance.AttendanceResponse;
import com.hrms.hr_payroll_management_system.dto.response.attendance.AttendanceMonthlySummaryResponse;
import com.hrms.hr_payroll_management_system.common.pagination.PageResponse;
import com.hrms.hr_payroll_management_system.enums.AttendanceStatus;
import java.time.LocalDate;

public interface AttendanceService {

    AttendanceResponse checkIn(Long employeeId);

    AttendanceResponse checkOut(Long employeeId);

    // Self-service — resolves the employee record from the logged-in
    // user's email, so employees can only ever punch their own record.
    AttendanceResponse checkInSelf(String email);

    AttendanceResponse checkOutSelf(String email);

    AttendanceResponse getTodayForSelf(String email); // null if not checked in yet

    AttendanceResponse getById(Long id);

    AttendanceResponse adjust(
        Long attendanceId,
        AdjustAttendanceRequest request
    );

    PageResponse<AttendanceResponse> search(
        int page,
        int size,
        Long employeeId,
        Long departmentId,
        Long branchId,
        AttendanceStatus status,
        LocalDate startDate,
        LocalDate endDate
);

AttendanceMonthlySummaryResponse getMonthlySummary(
        Long employeeId,
        int year,
        int month
);
    AttendanceResponse scanQr(String email, String token);
}

File: frontend/src/services/attendance.service.js
import { attendanceApi } from '../api/attendance.api';

export const attendanceService = {
  // Self-service
  async getTodaySelf() {
    const res = await attendanceApi.getTodaySelf();
    return res.data; // null if not checked in today
  },
  async checkInSelf() {
    const res = await attendanceApi.checkInSelf();
    return res.data;
  },
  async checkOutSelf() {
    const res = await attendanceApi.checkOutSelf();
    return res.data;
  },

  // Admin/HR
  async search({ page = 0, size = 10, employeeId, departmentId, branchId, status, startDate, endDate } = {}) {
    const res = await attendanceApi.search({ page, size, employeeId, departmentId, branchId, status, startDate, endDate });
    return res.data;
  },
  async getMonthlySummary(employeeId, year, month) {
    const res = await attendanceApi.getMonthlySummary(employeeId, year, month);
    return res.data;
  },
  async adjust(id, payload) {
    const res = await attendanceApi.adjust(id, payload);
    return res.data;
  },
};